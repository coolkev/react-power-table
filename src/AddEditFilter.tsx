import * as React from 'react';
import * as Button from 'react-bootstrap/lib/Button';
import * as Radio from 'react-bootstrap/lib/Radio';
import * as filters from './filters/FilterDefinition';
import { debuglog } from './utils';

/**
 * @internal
 */
export interface AddEditFilterProps {
    filter: filters.FilterDefinition<any>;
    initialOperation: filters.OperationDefinition<any>;
    initialValue?: any;
    onApplyFilter: (filter: filters.AppliedFilter<any>) => void;
}

/**
 * @internal
 */
export interface AddEditFilterState {

    operationKey: string;
    value: any;

}

/**
 * @internal
 */
export class AddEditFilter extends React.PureComponent<AddEditFilterProps, AddEditFilterState> {

    constructor(props: AddEditFilterProps) {
        super(props);

        const value = props.initialValue || props.filter.defaultValue || '';

        this.state = { value, operationKey: props.initialOperation.key };

        this.selectedOperationChange = this.selectedOperationChange.bind(this);
        this.handleFilterValueChange = this.handleFilterValueChange.bind(this);
        this.applyFilter = this.applyFilter.bind(this);

    }

    private selectedOperationChange(e: React.FormEvent<Radio & HTMLInputElement>) {

        const prevOperationKey = this.state.operationKey;
        const operationKey = e.currentTarget.value;
        //const operation = this.props.filter.operations[operationKey];

        const newState: AddEditFilterState = { operationKey, value: this.state.value };

        this.props.filter.onOperationChange(prevOperationKey, newState);
        // if (prevOperationKey === 'between' && operationKey !== prevOperationKey && typeof this.state.value === 'string') {
        //     newState.value = this.state.value.split(' ')[0];
        // }
        this.setState(newState);
    }

    private handleFilterValueChange(value: string) {
        debuglog('AddEditFilter.filterValueChange', value);
        this.setState({
            value,
        });
    }

    private applyFilter() {
        const operation = this.props.filter.operations[this.state.operationKey];

        this.props.onApplyFilter({ filter: this.props.filter, operation, value: this.state.value });
    }
    render() {
        debuglog('AddEditFilter.render', this.props);

        const filter = this.props.filter;

        const { value, operationKey } = this.state;

        const opComponents = Object.keys(filter.operations).map((opKey) => {
            const op = filter.operations[opKey];
            let children: React.ReactChild;
            if (opKey === this.state.operationKey) {

                const SelectedFilterComponent = op.filterComponent || filter.filterComponent;

                if (SelectedFilterComponent) {
                    children = <div style={{ marginLeft: 20 }}><SelectedFilterComponent filter={filter} operation={op} value={value} onValueChange={this.handleFilterValueChange} onEnterKeyPress={this.applyFilter} /></div>;
                }
            }

            return <FilterOperation key={opKey} operation={op} filter={filter} selected={opKey === operationKey} onChange={this.selectedOperationChange}>{children}</FilterOperation>;
        });

        return (
            <div>
                <div style={{ margin: '10px 0' }}><b>{filter.displayName}</b>
                </div>

                {opComponents.map((opComponent) => opComponent)}

                <div style={{ marginTop: 20 }}>
                    <Button bsStyle="primary" bsSize="sm" onClick={this.applyFilter}>Apply Filter</Button>
                </div>

            </div>);
    }
}

/**
 * @internal
 */
export interface FilterOperationProps extends React.Props<any> {
    operation: filters.OperationDefinition;
    filter: filters.FilterDefinition;

    selected: boolean;
    onChange: React.EventHandler<React.FormEvent<Radio & HTMLInputElement>>;
}

/**
 * @internal
 */
export const FilterOperation = (props: FilterOperationProps) => {

    const { operation, filter, onChange, selected, children } = props;

    return (
        <div key={operation.key}>
            <Radio checked={selected} onChange={onChange} value={operation.key}>
                {filter.radioButtonLabel ? filter.radioButtonLabel({ filter, operation }) : operation.displayName}

            </Radio>
            {children}
        </div>
    );

};
