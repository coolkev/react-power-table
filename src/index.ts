export * from './ReactPowerTable';
export * from './Paging';
export * from './Sorting';
export * from './GridFilters';
export * from './filters/';
export * from './Column';
export * from './utils';

import { PowerTableProps } from './ReactPowerTable';
export type ReactPowerTable<TRow = {}, TExtraProps = {}> = React.ComponentClass<PowerTableProps<TRow, TExtraProps>>;
