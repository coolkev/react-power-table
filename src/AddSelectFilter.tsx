import * as React from 'react';
import * as FormControl from 'react-bootstrap/lib/FormControl';
import { AddEditFilter } from './AddEditFilter';
import { BackLink } from './components/BackLink';
import * as PowerTable from './filters/FilterDefinition';
import { objectMapToArray } from './utils';

/**
 * @internal
 */
export interface AddFilterProps {
    cancelAddFilter: () => void;

    availableFilters: PowerTable.FilterDefinition[] | { [key: string]: PowerTable.FilterDefinition };

    //availableFilters: { [key: string]: PowerTable.FilterDefinition };
    appliedFilters: PowerTable.AppliedFilter[];
    onApplyFilter: (filter: PowerTable.AppliedFilter) => void;

}

/**
 * @internal
 */
export interface AddFilterState {
    searchText: string;
    selectedFilterKey?: string;
}

/**
 * @internal
 */
export class AddSelectFilter extends React.PureComponent<AddFilterProps, AddFilterState> {

    constructor(props: AddFilterProps) {
        super(props);
        this.state = { searchText: '' };
        this.onSearchTextChanged = this.onSearchTextChanged.bind(this);
        this.backToPrev = this.backToPrev.bind(this);
        this.newFilterSelected = this.newFilterSelected.bind(this);

    }

    private onSearchTextChanged(e: React.FormEvent<FormControl & HTMLInputElement>) {
        this.setState({
            searchText: e.currentTarget.value,
        });

    }

    private newFilterSelected(e: React.SyntheticEvent<HTMLAnchorElement>) {
        e.preventDefault();

        this.setState({
            selectedFilterKey: e.currentTarget.dataset.fieldName,
        });
    }

    private backToPrev(e: React.SyntheticEvent<HTMLAnchorElement>) {
        e.preventDefault();

        if (this.state.selectedFilterKey) {
            this.setState({ selectedFilterKey: null });
        } else {
            this.props.cancelAddFilter();
        }
    }
    render() {

        const { availableFilters, appliedFilters } = this.props;

        const { searchText, selectedFilterKey } = this.state;

        if (selectedFilterKey) {

            const filter = Array.isArray(availableFilters) ? availableFilters.find((f) => f.fieldName === selectedFilterKey) : availableFilters[selectedFilterKey];

            const initialOperation = objectMapToArray(filter.operations)[0];

            return (
                <div>
                    <BackLink onClick={this.backToPrev} />

                    <div>

                        <AddEditFilter filter={filter} initialOperation={initialOperation} onApplyFilter={this.props.onApplyFilter} />

                    </div>
                </div>
            );

        }

        const filters = objectMapToArray(availableFilters);

        const unusedFilters = filters.filter((m) => appliedFilters.every((c) => c.filter.fieldName !== m.fieldName));

        const regex = new RegExp(searchText, 'i');
        const showFilters = searchText ? unusedFilters.filter((m) => m.displayName.match(regex)) : unusedFilters;

        // const onSelectFilter = (e: React.) => {
        //     e.preventDefault();
        //     this.newFilterSelected(m.fieldName);
        // };

        return (
            <div className="flex-column">
                <BackLink onClick={this.backToPrev} />

                <div><FormControl placeholder="Filter by" value={searchText} onChange={this.onSearchTextChanged} autoFocus /></div>

                <div className="small flex-column">
                    <div style={{ margin: '10px 0' }}><b>Available Filters</b></div>
                    <div className="available-filters">
                        <div className="list-group">
                            {showFilters.map((m) => <a href="#" onClick={this.newFilterSelected} data-fieldName={m.fieldName} className="list-group-item" key={m.fieldName}>{m.displayName}</a>)}
                        </div>
                    </div>

                </div>

            </div>
        );
    }
}
