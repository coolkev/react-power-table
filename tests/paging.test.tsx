import * as React from 'react';
import { ReactPowerTable, GridProps, withInternalPaging, InternalPagingProps } from '../src/';
import { defaultColumns, sampledata, President } from './shared';
import { mount } from 'enzyme';

const columns = defaultColumns;
const rows = sampledata.slice(0, 25);

describe('withInternalPaging tests',
    () => {

        test('render paging', () => {

            const Table = withInternalPaging(ReactPowerTable);

            const c = <Table columns={columns} rows={rows} keyColumn="number" />;
            const component = mount(c);

            expect(component.state('currentPage')).toEqual(1);

            expect(component.render()).toMatchSnapshot();

        });

        test('change currentPage from click', () => {

            const Table = withInternalPaging(ReactPowerTable);

            const c: React.ReactElement<GridProps<President> & { paging?: Partial<InternalPagingProps> }> = <Table columns={columns} rows={rows} keyColumn="number" />;

            const component = mount(c);
            expect(component.find('tbody').first().find('tr').length).toEqual(20);

            expect(component.state('currentPage')).toEqual(1);
            component.find('.glyphicon-forward').simulate('click');
            expect(component.state('currentPage')).toEqual(2);

            expect(component.find('tbody').first().find('tr').length).toEqual(5);
            expect(component.render()).toMatchSnapshot();

        });

        test('change currentPage from props', () => {

            const Table = withInternalPaging(ReactPowerTable);

            const c: React.ReactElement<GridProps<President> & { paging?: Partial<InternalPagingProps> }> = <Table columns={columns} rows={rows} keyColumn="number" />;

            const component = mount(c);
            expect(component.find('tbody').first().find('tr').length).toEqual(20);
            component.setProps({ paging: { currentPage: 2 } });

            expect(component.find('tbody').first().find('tr').length).toEqual(5);
            expect(component.render()).toMatchSnapshot();

        });

    });
