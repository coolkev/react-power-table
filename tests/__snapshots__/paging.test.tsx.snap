// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withInternalPaging tests change currentPage from click 1`] = `
<table>
  <thead>
    <tr>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          number
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          president
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          party
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          birth_year
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          death_year
        </div>
      </th>
      <th
        style="text-align: right; white-space: nowrap; width: 140px;"
      >
        <div>
          took_office
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          left_office
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <div>
          21
        </div>
      </td>
      <td>
        <div>
          Chester A. Arthur
        </div>
      </td>
      <td>
        <div>
          Republican
        </div>
      </td>
      <td>
        <div>
          1829
        </div>
      </td>
      <td>
        <div>
          1886
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1881-09-19
        </div>
      </td>
      <td>
        <div>
          1885-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          22
        </div>
      </td>
      <td>
        <div>
          Grover Cleveland
        </div>
      </td>
      <td>
        <div>
          Democratic
        </div>
      </td>
      <td>
        <div>
          1837
        </div>
      </td>
      <td>
        <div>
          1908
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1885-03-04
        </div>
      </td>
      <td>
        <div>
          1889-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          23
        </div>
      </td>
      <td>
        <div>
          Benjamin Harrison
        </div>
      </td>
      <td>
        <div>
          Republican
        </div>
      </td>
      <td>
        <div>
          1833
        </div>
      </td>
      <td>
        <div>
          1901
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1889-03-04
        </div>
      </td>
      <td>
        <div>
          1893-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          24
        </div>
      </td>
      <td>
        <div>
          Grover Cleveland
        </div>
      </td>
      <td>
        <div>
          Democratic
        </div>
      </td>
      <td>
        <div>
          1837
        </div>
      </td>
      <td>
        <div>
          1908
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1893-03-04
        </div>
      </td>
      <td>
        <div>
          1897-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          25
        </div>
      </td>
      <td>
        <div>
          William McKinley
        </div>
      </td>
      <td>
        <div>
          Republican
        </div>
      </td>
      <td>
        <div>
          1843
        </div>
      </td>
      <td>
        <div>
          1901
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1897-03-04
        </div>
      </td>
      <td>
        <div>
          1901-09-14
        </div>
      </td>
    </tr>
  </tbody>
  <tfoot>
    <tr>
      <td
        colspan="7"
      >
        <table
          class="form-inline"
          width="100%"
        >
          <tbody>
            <tr>
              <td
                style="font-weight: bold;"
                width="33%"
              >
                25 Records
              </td>
              <td
                style="text-align: center;"
                width="34%"
              >
                <a
                  class="glyphicon glyphicon-fast-backward"
                  href="#"
                  style="text-decoration: none;"
                />
                 
                <a
                  class="glyphicon glyphicon-backward"
                  href="#"
                  style="text-decoration: none;"
                />
                 
                <span>
                  Page  
                  <span>
                    <input
                      class="form-control input-sm"
                      style="width: 60px;"
                      type="text"
                      value="2"
                    />
                  </span>
                    of 
                  2
                </span>
                 
                <a
                  class="glyphicon glyphicon-forward"
                  href="#"
                  style="text-decoration: none; color: silver; cursor: default;"
                />
                 
                <a
                  class="glyphicon glyphicon-fast-forward"
                  href="#"
                  style="text-decoration: none; color: silver; cursor: default;"
                />
              </td>
              <td
                style="text-align: right;"
                width="33%"
              >
                Show:
                <select
                  class="form-control input-sm"
                  style="width: 80px;"
                >
                  <option
                    value="10"
                  >
                    10
                  </option>
                  <option
                    value="20"
                  >
                    20
                  </option>
                  <option
                    value="50"
                  >
                    50
                  </option>
                  <option
                    value="100"
                  >
                    100
                  </option>
                  <option
                    value="500"
                  >
                    500
                  </option>
                </select>
              </td>
            </tr>
          </tbody>
        </table>
      </td>
    </tr>
  </tfoot>
</table>
`;

exports[`withInternalPaging tests change currentPage from props 1`] = `
<table>
  <thead>
    <tr>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          number
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          president
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          party
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          birth_year
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          death_year
        </div>
      </th>
      <th
        style="text-align: right; white-space: nowrap; width: 140px;"
      >
        <div>
          took_office
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          left_office
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <div>
          21
        </div>
      </td>
      <td>
        <div>
          Chester A. Arthur
        </div>
      </td>
      <td>
        <div>
          Republican
        </div>
      </td>
      <td>
        <div>
          1829
        </div>
      </td>
      <td>
        <div>
          1886
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1881-09-19
        </div>
      </td>
      <td>
        <div>
          1885-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          22
        </div>
      </td>
      <td>
        <div>
          Grover Cleveland
        </div>
      </td>
      <td>
        <div>
          Democratic
        </div>
      </td>
      <td>
        <div>
          1837
        </div>
      </td>
      <td>
        <div>
          1908
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1885-03-04
        </div>
      </td>
      <td>
        <div>
          1889-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          23
        </div>
      </td>
      <td>
        <div>
          Benjamin Harrison
        </div>
      </td>
      <td>
        <div>
          Republican
        </div>
      </td>
      <td>
        <div>
          1833
        </div>
      </td>
      <td>
        <div>
          1901
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1889-03-04
        </div>
      </td>
      <td>
        <div>
          1893-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          24
        </div>
      </td>
      <td>
        <div>
          Grover Cleveland
        </div>
      </td>
      <td>
        <div>
          Democratic
        </div>
      </td>
      <td>
        <div>
          1837
        </div>
      </td>
      <td>
        <div>
          1908
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1893-03-04
        </div>
      </td>
      <td>
        <div>
          1897-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          25
        </div>
      </td>
      <td>
        <div>
          William McKinley
        </div>
      </td>
      <td>
        <div>
          Republican
        </div>
      </td>
      <td>
        <div>
          1843
        </div>
      </td>
      <td>
        <div>
          1901
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1897-03-04
        </div>
      </td>
      <td>
        <div>
          1901-09-14
        </div>
      </td>
    </tr>
  </tbody>
  <tfoot>
    <tr>
      <td
        colspan="7"
      >
        <table
          class="form-inline"
          width="100%"
        >
          <tbody>
            <tr>
              <td
                style="font-weight: bold;"
                width="33%"
              >
                25 Records
              </td>
              <td
                style="text-align: center;"
                width="34%"
              >
                <a
                  class="glyphicon glyphicon-fast-backward"
                  href="#"
                  style="text-decoration: none;"
                />
                 
                <a
                  class="glyphicon glyphicon-backward"
                  href="#"
                  style="text-decoration: none;"
                />
                 
                <span>
                  Page  
                  <span>
                    <input
                      class="form-control input-sm"
                      style="width: 60px;"
                      type="text"
                      value="2"
                    />
                  </span>
                    of 
                  2
                </span>
                 
                <a
                  class="glyphicon glyphicon-forward"
                  href="#"
                  style="text-decoration: none; color: silver; cursor: default;"
                />
                 
                <a
                  class="glyphicon glyphicon-fast-forward"
                  href="#"
                  style="text-decoration: none; color: silver; cursor: default;"
                />
              </td>
              <td
                style="text-align: right;"
                width="33%"
              >
                Show:
                <select
                  class="form-control input-sm"
                  style="width: 80px;"
                >
                  <option
                    value="10"
                  >
                    10
                  </option>
                  <option
                    value="20"
                  >
                    20
                  </option>
                  <option
                    value="50"
                  >
                    50
                  </option>
                  <option
                    value="100"
                  >
                    100
                  </option>
                  <option
                    value="500"
                  >
                    500
                  </option>
                </select>
              </td>
            </tr>
          </tbody>
        </table>
      </td>
    </tr>
  </tfoot>
</table>
`;

exports[`withInternalPaging tests render paging 1`] = `
<table>
  <thead>
    <tr>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          number
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          president
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          party
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          birth_year
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          death_year
        </div>
      </th>
      <th
        style="text-align: right; white-space: nowrap; width: 140px;"
      >
        <div>
          took_office
        </div>
      </th>
      <th
        style="text-align: left; white-space: nowrap;"
      >
        <div>
          left_office
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <div>
          1
        </div>
      </td>
      <td>
        <div>
          George Washington
        </div>
      </td>
      <td>
        <div>
          No Party
        </div>
      </td>
      <td>
        <div>
          1732
        </div>
      </td>
      <td>
        <div>
          1799
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1789-04-30
        </div>
      </td>
      <td>
        <div>
          1797-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          2
        </div>
      </td>
      <td>
        <div>
          John Adams
        </div>
      </td>
      <td>
        <div>
          Federalist
        </div>
      </td>
      <td>
        <div>
          1735
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1797-03-04
        </div>
      </td>
      <td>
        <div>
          1801-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          3
        </div>
      </td>
      <td>
        <div>
          Thomas Jefferson
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1743
        </div>
      </td>
      <td>
        <div>
          1826
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1801-03-04
        </div>
      </td>
      <td>
        <div>
          1809-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          4
        </div>
      </td>
      <td>
        <div>
          James Madison
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1751
        </div>
      </td>
      <td>
        <div>
          1836
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1809-03-04
        </div>
      </td>
      <td>
        <div>
          1817-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          5
        </div>
      </td>
      <td>
        <div>
          James Monroe
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1758
        </div>
      </td>
      <td>
        <div>
          1831
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1817-03-04
        </div>
      </td>
      <td>
        <div>
          1825-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          6
        </div>
      </td>
      <td>
        <div>
          John Quincy Adams
        </div>
      </td>
      <td>
        <div>
          Democratic-Republican
        </div>
      </td>
      <td>
        <div>
          1767
        </div>
      </td>
      <td>
        <div>
          1848
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1825-03-04
        </div>
      </td>
      <td>
        <div>
          1829-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          7
        </div>
      </td>
      <td>
        <div>
          Andrew Jackson
        </div>
      </td>
      <td>
        <div>
          Democratic
        </div>
      </td>
      <td>
        <div>
          1767
        </div>
      </td>
      <td>
        <div>
          1845
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1829-03-04
        </div>
      </td>
      <td>
        <div>
          1837-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          8
        </div>
      </td>
      <td>
        <div>
          Martin Van Buren
        </div>
      </td>
      <td>
        <div>
          Democratic
        </div>
      </td>
      <td>
        <div>
          1782
        </div>
      </td>
      <td>
        <div>
          1862
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1837-03-04
        </div>
      </td>
      <td>
        <div>
          1841-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          9
        </div>
      </td>
      <td>
        <div>
          William Henry Harrison
        </div>
      </td>
      <td>
        <div>
          Whig
        </div>
      </td>
      <td>
        <div>
          1773
        </div>
      </td>
      <td>
        <div>
          1841
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1841-03-04
        </div>
      </td>
      <td>
        <div>
          1841-04-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          10
        </div>
      </td>
      <td>
        <div>
          John Tyler
        </div>
      </td>
      <td>
        <div>
          Whig
        </div>
      </td>
      <td>
        <div>
          1790
        </div>
      </td>
      <td>
        <div>
          1862
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1841-04-04
        </div>
      </td>
      <td>
        <div>
          1845-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          11
        </div>
      </td>
      <td>
        <div>
          James K. Polk
        </div>
      </td>
      <td>
        <div>
          Democratic
        </div>
      </td>
      <td>
        <div>
          1795
        </div>
      </td>
      <td>
        <div>
          1849
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1845-03-04
        </div>
      </td>
      <td>
        <div>
          1849-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          12
        </div>
      </td>
      <td>
        <div>
          Zachary Taylor
        </div>
      </td>
      <td>
        <div>
          Whig
        </div>
      </td>
      <td>
        <div>
          1784
        </div>
      </td>
      <td>
        <div>
          1850
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1849-03-04
        </div>
      </td>
      <td>
        <div>
          1850-07-09
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          13
        </div>
      </td>
      <td>
        <div>
          Millard Fillmore
        </div>
      </td>
      <td>
        <div>
          Whig
        </div>
      </td>
      <td>
        <div>
          1800
        </div>
      </td>
      <td>
        <div>
          1874
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1850-07-09
        </div>
      </td>
      <td>
        <div>
          1853-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          14
        </div>
      </td>
      <td>
        <div>
          Franklin Pierce
        </div>
      </td>
      <td>
        <div>
          Democratic
        </div>
      </td>
      <td>
        <div>
          1804
        </div>
      </td>
      <td>
        <div>
          1869
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1853-03-04
        </div>
      </td>
      <td>
        <div>
          1857-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          15
        </div>
      </td>
      <td>
        <div>
          James Buchanan
        </div>
      </td>
      <td>
        <div>
          Democratic
        </div>
      </td>
      <td>
        <div>
          1791
        </div>
      </td>
      <td>
        <div>
          1868
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1857-03-04
        </div>
      </td>
      <td>
        <div>
          1861-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          16
        </div>
      </td>
      <td>
        <div>
          Abraham Lincoln
        </div>
      </td>
      <td>
        <div>
          Republican
        </div>
      </td>
      <td>
        <div>
          1809
        </div>
      </td>
      <td>
        <div>
          1865
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1861-03-04
        </div>
      </td>
      <td>
        <div>
          1865-04-15
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          17
        </div>
      </td>
      <td>
        <div>
          Andrew Johnson
        </div>
      </td>
      <td>
        <div>
          Democratic
        </div>
      </td>
      <td>
        <div>
          1808
        </div>
      </td>
      <td>
        <div>
          1875
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1865-04-15
        </div>
      </td>
      <td>
        <div>
          1869-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          18
        </div>
      </td>
      <td>
        <div>
          Ulysses S. Grant
        </div>
      </td>
      <td>
        <div>
          Republican
        </div>
      </td>
      <td>
        <div>
          1822
        </div>
      </td>
      <td>
        <div>
          1885
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1869-03-04
        </div>
      </td>
      <td>
        <div>
          1877-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          19
        </div>
      </td>
      <td>
        <div>
          Rutherford B. Hayes
        </div>
      </td>
      <td>
        <div>
          Republican
        </div>
      </td>
      <td>
        <div>
          1822
        </div>
      </td>
      <td>
        <div>
          1893
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1877-03-04
        </div>
      </td>
      <td>
        <div>
          1881-03-04
        </div>
      </td>
    </tr>
    <tr>
      <td>
        <div>
          20
        </div>
      </td>
      <td>
        <div>
          James A. Garfield
        </div>
      </td>
      <td>
        <div>
          Republican
        </div>
      </td>
      <td>
        <div>
          1831
        </div>
      </td>
      <td>
        <div>
          1881
        </div>
      </td>
      <td
        style="width: 140px; text-align: right;"
      >
        <div>
          1881-03-04
        </div>
      </td>
      <td>
        <div>
          1881-09-19
        </div>
      </td>
    </tr>
  </tbody>
  <tfoot>
    <tr>
      <td
        colspan="7"
      >
        <table
          class="form-inline"
          width="100%"
        >
          <tbody>
            <tr>
              <td
                style="font-weight: bold;"
                width="33%"
              >
                25 Records
              </td>
              <td
                style="text-align: center;"
                width="34%"
              >
                <a
                  class="glyphicon glyphicon-fast-backward"
                  href="#"
                  style="text-decoration: none; color: silver; cursor: default;"
                />
                 
                <a
                  class="glyphicon glyphicon-backward"
                  href="#"
                  style="text-decoration: none; color: silver; cursor: default;"
                />
                 
                <span>
                  Page  
                  <span>
                    <input
                      class="form-control input-sm"
                      style="width: 60px;"
                      type="text"
                      value="1"
                    />
                  </span>
                    of 
                  2
                </span>
                 
                <a
                  class="glyphicon glyphicon-forward"
                  href="#"
                  style="text-decoration: none;"
                />
                 
                <a
                  class="glyphicon glyphicon-fast-forward"
                  href="#"
                  style="text-decoration: none;"
                />
              </td>
              <td
                style="text-align: right;"
                width="33%"
              >
                Show:
                <select
                  class="form-control input-sm"
                  style="width: 80px;"
                >
                  <option
                    value="10"
                  >
                    10
                  </option>
                  <option
                    value="20"
                  >
                    20
                  </option>
                  <option
                    value="50"
                  >
                    50
                  </option>
                  <option
                    value="100"
                  >
                    100
                  </option>
                  <option
                    value="500"
                  >
                    500
                  </option>
                </select>
              </td>
            </tr>
          </tbody>
        </table>
      </td>
    </tr>
  </tfoot>
</table>
`;
